### PR TITLE
Feature/trade manager

### PR DESCRIPTION
**Pull request by nomis51**
## Description
Add a trade manager. 
Handles incoming trade offers and show them directly on the overlay.
_The app can highlight items using two methods: in-game search and a grid component in the overlay (MercuryTrade style)_
(Outgoing offers ~~and item highlighting are~~ is in progress)

#### ChatListener
Looks for chat inputs in the Client.txt log file of PoE.

#### Trade Service
Handle interactions with the main process (ChatListener)

#### Trade Offers Container
Component that handle trade offers.

#### Trade Offer
Component that display the offer on screen (The tiles).

#### Trade settings
- Customize whispers with internal variables, such as the "item name" and the "price"
- Enable/Disable auto-kick and auto-whisper
- Enable/Disable item highlighting methods

_#### Trade Grid
It's the component used to highlight items on a grid using the overlay_ 

## Screenshots/Video
See the documentation about it. [Doc](https://github.com/nomis51/PoE-Overlay-Community-Fork/blob/feature/trade-manager/TRADE-MANAGER.md)

## How Has This Been Tested?
- [x] ran `npm run ng:lint` (There was existing lint errors not related to the new code)
- [x] ran `npm run format`
- [x] ran `npm run ng:test`
- [ ] added unit tests
- [x] manually tested (In dev, in prod and while playing the game)

## Aditionnal infos
- Add the **lodash** library
- Tested on Windows
- Tested with Vulkan and DX11

In Dev mode, when the Hot Reload module of Angular trigger, the app loose the reference to the function that enable/disable mouse events, so you need to restart the electron part of the app to re-enable that function.

### CODE REVIEW

**Comment by WhiteFang5** [comment]
Perhaps you could look into completing implementation for the outgoing trades and add support for quad-tabs?

File: electron/game.ts [hunk 1]
```diff
@@ -23,21 +23,30 @@ const POE_TITLES = ['Path of Exile']
...
 
     const window = await getActiveWindow()
     if (window) {
-      const windowPath = (window.path || '').toLowerCase()
+      const windowPath = (window.path || '').toLowerCase();
       const name = path.basename(windowPath)
       if (POE_NAMES.includes(name)) {
         const title = window.title()
         if (POE_TITLES.includes(title) || POE_ALTERNATIVE_TITLES.some((x) => title.startsWith(x))) {
+        this.ipcMain.emit('logs-file-found', `${windowPath.substring(0, windowPath.lastIndexOf('\\'))}\\logs\\Client.txt`);
```
Comment (Bubblegnash) [inline electron/game.ts:48]:
```diff
electron/game.ts

-this.ipcMain.emit('logs-file-found', `${windowPath.substring(0, windowPath.lastIndexOf('\\'))}\\logs\\Client.txt`);
+this.ipcMain.emit('logs-file-found', path.join(path.parse(windowPath).dir, "logs" , "Client.txt")
```
so you don't break mac/Linux compatibility

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+## 0.7.3 (2020-07-04)
```
Comment (versedi) [inline CHANGELOG.md:3]: 0.8.0, major feature, see https://semver.org/

File: src/app/modules/trade/components/trade-offers-container/trade-offers-container.component.html [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  <app-trade-stash-grid [location]="currentOffer ? currentOffer.itemLocation : {left:0, top: 0, width: 0, height: 0}" [dropShadow]="dropShadow" [darkerShadow]="darkerShadow" [demo]="gridDemo" [gridLocation]="gridLocation" (evTop)="updateGridPosition('top', $event)" (evLeft)="updateGridPosition('left', $event)"></app-trade-stash-grid>
+</div>
+
+<button class="btn-adjust-grid" mat-raised-button *ngIf="showGrid" (click)="setGridDemo(!currentOffer)">{{gridDemo ? 'Save Grid' : 'Adjust Grid'}}</button>
+
+<mat-icon class="icon-adjust-grid" (click)="setGridDemo(true);">grid_on</mat-icon>
+
+<p class="lbl-adjust-message" *ngIf="gridDemo && showGrid">Please adjust the grid approximately to your stash tab layout. <br>You need to do that, because every player use different screen resolution, size and settings. <br>But, don't worry you'll have to do this only once.</p>
+
+<div *ngIf="currentOffer" class="item-location">
```
Comment (versedi) [inline src/app/modules/trade/components/trade-offers-container/trade-offers-container.component.html:29]: This needs a display property, after doing a trade and going back into a map I'm stuck with: 

display property needed

File: electron/chat-listener.ts [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+                });
+            });
+        });
+    }
+
+    private parseOfferLine(line: string): void {
+        var buyerName = line.match(regGuild) ? line.substring(line.indexOf('> ') + 2, line.lastIndexOf(' Hi, ')) : line.substring(line.indexOf('@From ') + 6, line.match(regTwoDots) ? line.lastIndexOf(': Hi, ') : line.lastIndexOf(' Hi, '));
+
+        if (buyerName.lastIndexOf(':') == buyerName.length) {
+            buyerName = buyerName.substring(0, buyerName.length - 1);
```
Comment (versedi) [inline electron/chat-listener.ts:816]: **Bug**
### Invite:

Invite

### Chat input result:
InputResult

```
             buyerName = buyerName.replace(':', ''); 

```

Does it's job - you cant have a colon `:` in a character name. 



**Bug 2**
Lack of catching errors / fallback.
 
```
zone.js:192 Uncaught TypeError: Cannot read property 'buyerName' of undefined
    at SafeSubscriber._next (trade-offers-container.component.ts:226)
    at SafeSubscriber.push../node_modules/rxjs/_esm5/internal/Subscriber.js.SafeSubscriber.__tryOrUnsub (Subscriber.js:192)
    at SafeSubscriber.push../node_modules/rxjs/_esm5/internal/Subscriber.js.SafeSubscriber.next (Subscriber.js:130)
    at Subscriber.push../node_modules/rxjs/_esm5/internal/Subscriber.js.Subscriber._next (Subscriber.js:76)
    at Subscriber.push../node_modules/rxjs/_esm5/internal/Subscriber.js.Subscriber.next (Subscriber.js:53)
    at MergeMapSubscriber.push../node_modules/rxjs/_esm5/internal/operators/mergeMap.js.MergeMapSubscriber.notifyNext (mergeMap.js:87)
    at InnerSubscriber.push../node_modules/rxjs/_esm5/internal/InnerSubscriber.js.InnerSubscriber._next (InnerSubscriber.js:15)
    at InnerSubscriber.push../node_modules/rxjs/_esm5/internal/Subscriber.js.Subscriber.next (Subscriber.js:53)
    at Observable._subscribe (subscribeToArray.js:5)
    at Observable.push../node_modules/rxjs/_esm5/internal/Observable.js.Observable._trySubscribe (Observable.js:43)
```

File: TRADE-MANAGER.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+You'll have now, several actions available:
+- Send a trade request to the buyer (to engage the trade in-game)
+- Invite the buyer to your party (Usefull if the buyer, didn't accept your first party invite quick enough and needs another one.)
+- Kick the buyer out of your party (Usefull when the game prevent the buyer from joining your hideout with the "Failed to join instance" bug. You can then Kick the buyer and invite him again quickly)
+- Remove the offer (Removes the trade offer from the overlay **and** kick the buyer out of your party)
+
+![image](https://user-images.githubusercontent.com/25111613/86523398-d4279a80-be39-11ea-8af0-c9392eb399df.png)
+
+Once the trade is successful, the buyer will be automatically kicked out of your party and a "Thanks" whisper will be sent to him. (The trade offer is removed from the overlay aswell)
```
Comment (versedi) [inline TRADE-MANAGER.md:47]: > and a "Thanks" whisper will be sent to him. (The trade offer is removed from the overlay aswell)

Bugged, doesn't work. Kicked, no thank you. 
 Kick, no thanks :(


Works if you disable Auto-Kick in settings. Seems it doesn't wait for the message to be sent before kicking the trader out. 

Without auto-kick

**Comment by versedi** [comment]
@nomis51 I've resolved the conflicts locally to build it, if you want I can push it into my fork so you will have less things to do.

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "poe-overlay",
-  "version": "0.7.2",
+  "version": "0.7.3",
```
Comment (versedi) [inline package.json:3]:
```
   "version": "0.8.0", 
```

**Comment by versedi** [comment]
README.md needs updating - there's no info about trade manager nor link to it, see: https://github.com/versedi/PoE-Overlay-Community-Fork/commit/4d0d2ad7256276bcf6f1356ed6879ca0da8bb9bc

Updated in my fork

**Comment by Barry3D** [comment]
> Hi, I have no idea how to use Github and I'm mildly stoned, but does this somehow let me do trades via poe-overlay community fork?  I uninstalled the overwolf version b/c I didn't like it but I see the new version doesn't have a trade manager like it does.   If it does let me do those trades...how the fuck do I install it? ;)  Thanks!

This is an open feature branch. It will be merged and released once it's completed. As far as I can see it is more of a utility to allow faster trading and not an auction house.

If you want to install it before it is released you will have to finish the dev-setup described in the docs and check this branch out.

**Comment by nomis51** [comment]
It's a trade manager tool built in PoE Overlay, like Poe-Lurker and MercuryTrade. @versedi is, I guess, working on it, because i'm too busy right now to continue the developement.

**Comment by versedi** [comment]
> It's a trade manager tool built in PoE Overlay, like Poe-Lurker and MercuryTrade. @versedi is, I guess, working on it, because i'm too busy right now to continue the developement.

I've tested things out and there's couple of bugs I'd like to get rid off as top priority.

I've seen you re-created the repository on your profile, should I start off from there or here?

**Comment by WhiteFang5** [comment]
Closing this PR in favour of #200

**Pull request closed without merging**